### PR TITLE
Implement a data validation after uploading

### PR DESCRIPTION
**Pull request by tiwalter**
- Implement the FSReleaseArtifactService function ValidateUploadPayload()
- Implement an extension method that validates the DeploymentMetaInfo
- Extend the TestUtils
- Extend the DeploymentMetaInfoMapper to support a mapping of byte arrays
- Write unit tests for the new implementations
- Rename the former "Common" namespaces to "Extensions"
- Add test data for the new unit tests

### CODE REVIEW

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/ByteArrayValidationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+using System.Text;
+using Newtonsoft.Json;
+
+namespace ReleaseServer.WebApi.Extensions
+{
+    public static class ByteArrayValidationExtensions
+    {
+        public static bool IsAValidJson(this byte[] byteArray, out string exceptionMessage)
```
Comment (f-porter) [inline Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/ByteArrayValidationExtensions.cs:8]: Using a method like that is not good practice as you are deserializing an object that will be deserialized anyway in a later step if it is valid JSON.
So in the case of valid JSON the deserialization will be done twice.

When you are deserializing the object to e.g. `DeploymentInfoModel` it will also be checked if the the JSON is valid so there is no need for a check beforehand.

File: Sources/ReleaseServer/ReleaseServer.WebApi/Models/DeploymentMetaInfoModel.cs [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 namespace ReleaseServer.WebApi.Config
 {
-    public class DeploymentMetaInfo
+    public class DeploymentMetaInfoModel
```
Comment (f-porter) [inline Sources/ReleaseServer/ReleaseServer.WebApi/Models/DeploymentMetaInfoModel.cs:5]: Why did you add the suffix "Model". IMHO this is only necessary/useful, if there also is a class `DeploymentMetaInfo` and this is the corresponding model for that.

File: Sources/ReleaseServer/ReleaseServer.WebApi/Models/ValidationResultModel.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+namespace ReleaseServer.WebApi.Models
+{
+    public class ValidationResultModel
```
Comment (f-porter) [inline Sources/ReleaseServer/ReleaseServer.WebApi/Models/ValidationResultModel.cs:3]: Same as above: If there is no other class `ValidationResult`, why do you need a "Model" of it?

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        {
+            if (deploymentMetaInfo.ReleaseDate == new DateTime() || deploymentMetaInfo.ArtifactFileName == null ||
+                deploymentMetaInfo.ChangelogFileName == null)
+            {
+                return false;
+            }
+
+            return true;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs:18]: Done, see: https://github.com/Traeger-GmbH/release-server/commit/e15657b6efe1c11f51156d77084e26f90047c08f

File: Sources/ReleaseServer/ReleaseServer.WebApi/Services/FsReleaseArtifactService.cs [hunk 5]
```diff
@@ -170,5 +173,65 @@ public async Task RestoreBackup(IFormFile payload)
                 }
             }
         }
+
+        public async Task<ValidationResultModel> ValidateUploadPayload(IFormFile payload)
+        {
+            using (var zipMapper = new ZipArchiveMapper())
```
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Services/FsReleaseArtifactService.cs]: Done, see https://github.com/Traeger-GmbH/release-server/commit/e15657b6efe1c11f51156d77084e26f90047c08f

File: Sources/ReleaseServer/ReleaseServer.WebApi/Common/StreamToByteArray.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+using System.IO;
+using System.Threading.Tasks;
+
+namespace ReleaseServer.WebApi.Common
```
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Common/StreamToByteArray.cs:4]: I eliminated this function, because it is not needed anymore. The usage of a `StreamReader` made it obsolete.

File: Sources/ReleaseServer/ReleaseServer.WebApi/Common/StreamToByteArray.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+{
+    public static class StreamToByteArray
```
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Common/StreamToByteArray.cs:6]: The function got eliminated (see comment above)

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
+using System;
+using ReleaseServer.WebApi.Config;
+
+namespace ReleaseServer.WebApi.Extensions
+{
+    public static class DeploymentMetaInfoExtension
+    {
+        public static bool IsValid(this DeploymentMetaInfoModel deploymentMetaInfoModel)
```
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs]: I eliminated this function incl. the obsolete unit tests. 

Regarding the tip with `Nullable<DateTime>`: I tested deserializing a deployment.jso without a `ReleaseDate` field and with the `ReleaseDate` field wit an empty string value.
The deserializer recognized, that the `ReleaseDate` was not set.

Because of that, i didn't change this property to a `Nullable`.

Do you agree or did i misunderstand your intention?

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/ByteArrayValidationExtensions.cs (hunk 1 above)
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/ByteArrayValidationExtensions.cs:8]: I eliminated this function, because it's not needed anymore.

File: Sources/ReleaseServer/ReleaseServer.WebApi/Models/DeploymentMetaInfoModel.cs (hunk 2 above)
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Models/DeploymentMetaInfoModel.cs]: I removed the "Model" suffix, see https://github.com/Traeger-GmbH/release-server/commit/014b059c85cec990a6c6c065dd406057b121f649

File: Sources/ReleaseServer/ReleaseServer.WebApi/Models/ValidationResultModel.cs (hunk 3 above)
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Models/ValidationResultModel.cs]: I removed the "Model" suffix, see https://github.com/Traeger-GmbH/release-server/commit/014b059c85cec990a6c6c065dd406057b121f649

File: Sources/ReleaseServer/ReleaseServer.WebApi/Services/FsReleaseArtifactService.cs [hunk 9]
```diff
@@ -170,5 +173,69 @@ public async Task RestoreBackup(IFormFile payload)
...
+                if (deploymentInfoEntry == null)
+                {
+                    var validationError = "the deployment.json does not exist in the uploaded payload!";
+                    Logger.LogError(validationError);
+                    return new ValidationResultModel {IsValid = false, ValidationError = validationError};
+                }
+
+                //Open the deployment.json and extract the DeploymentMetaInfo of it
+                var deploymentInfoStream = deploymentInfoEntry.Open();
+                var deploymentInfoByteArray = await deploymentInfoStream.ToByteArrayAsync();
```
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Services/FsReleaseArtifactService.cs]: Done, see: https://github.com/Traeger-GmbH/release-server/commit/3f9d9fb070b98c88cbe9b747abc28b7f2aee3e72

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/ByteArrayValidationExtensions.cs (hunk 1 above)
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/ByteArrayValidationExtensions.cs]: See: https://github.com/Traeger-GmbH/release-server/commit/3f9d9fb070b98c88cbe9b747abc28b7f2aee3e72

File: Sources/ReleaseServer/ReleaseServer.WebApi/Common/StreamToByteArray.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Common/StreamToByteArray.cs]: See https://github.com/Traeger-GmbH/release-server/commit/3f9d9fb070b98c88cbe9b747abc28b7f2aee3e72

File: Sources/ReleaseServer/ReleaseServer.WebApi/Common/StreamToByteArray.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Common/StreamToByteArray.cs]: See https://github.com/Traeger-GmbH/release-server/commit/3f9d9fb070b98c88cbe9b747abc28b7f2aee3e72

**Comment by tiwalter** [comment]
@f-porter 

> 
> 
> After reconsidering the tests using zip archives that have been implemented here:
> 
> I understand the necessity of testing the functionality from end to end.
> But actually it is not feasible to maintain tests that are based on these zip files. In case the interface has to be changed and the tests have to be adapted according to these changes one has to open every single one of these archives, edit the contained files and replace the checked in binary file by the new zip archive.
> In addition to that it is not good practice to add binary files in a git repository if one can avoid it. This unnecessarily increases the repository size with every change done on such a file.
> 
> I would suggest adding just one zip archive to ensure a correct access on a real zip file and for the other tests you could for example programmatically generate the zip files as part of the test setup.

i will refactor the unit testing in the second PR: https://github.com/Traeger-GmbH/release-server/pull/59 to avoid merge conflicts.

**Comment by tiwalter** [comment]
Refactoring of the upload validation testing done here: https://github.com/Traeger-GmbH/release-server/commit/35d0369df10f163dce6be2521134a325d67dd0e9

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (f-porter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs]: Where do I find the unit test for this?

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (f-porter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs]: Actually I do not find the validation of the object in the commit your referenced...

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs]: I manually tested it, if you want, i can add a unit test for this cases

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs]: The validation will be done automatically during deserialization (check if it is a valid json and if the required properties are set). See the unit tests in the other PR: https://github.com/Traeger-GmbH/release-server/blob/feature/changelog-with-categories/Sources/ReleaseServer/ReleaseServer.WebApi.Test/Services/ReleaseArtifactServiceTest.cs#L137

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (f-porter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs]: Yes, things like correct serialization, deserialization and detection of invalid data **must** be covered by respective unit tests.

In general:
In the case you write any test (like this one) for manual testing - just create a unit test for that as this usually does not take more effort. Otherwise this is a waste of work.

[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs]: Okay, i can add unit tests for this.

In my opinion, it didn't make much sense to test the functionality of `[Newtonsoft.Json.JsonRequired]` in an extra unit test, because will be already tested in their project.

File: Sources/ReleaseServer/ReleaseServer.WebApi/Mappers/DeploymentMetaMapper.cs [hunk 10]
```diff
@@ -1,14 +1,20 @@
...
-        public static DeploymentMetaInfo ParseDeploymentMetaInfo(string fileName)
+        public static DeploymentMetaInfoModel ParseDeploymentMetaInfo(string fileName)
         {
-            return JsonConvert.DeserializeObject<DeploymentMetaInfo>(File.ReadAllText(fileName));
+            return JsonConvert.DeserializeObject<DeploymentMetaInfoModel>(File.ReadAllText(fileName));
+        }
+        
+        public static DeploymentMetaInfoModel ParseDeploymentMetaInfo(byte[] byteContent)
+        {
+            return JsonConvert.DeserializeObject<DeploymentMetaInfoModel>(Encoding.UTF8.GetString(byteContent));
```
Comment (f-porter) [inline Sources/ReleaseServer/ReleaseServer.WebApi/Mappers/DeploymentMetaMapper.cs]: This method is only used in the test project. So it seems to be obsolete.

File: Sources/ReleaseServer/ReleaseServer.WebApi/Services/FsReleaseArtifactService.cs [hunk 11]
```diff
@@ -152,23 +155,80 @@ public async Task<BackupInformationModel> RunBackup()
...
-                //It's important to release the semaphore. try / finally block ensures a guaranteed release (also if the operation may crash) 
+            //Open the deployment.json and extract the DeploymentMetaInfo of it
+            using (StreamReader deploymentInfoFile = new StreamReader(deploymentInfoEntry.Open(), System.Text.Encoding.UTF8))
+            {
                 try
                 {
-                    await Task.Run(() => FsReleaseArtifactRepository.RestoreBackup(zipPayload));
+                    JsonSerializer serializer = new JsonSerializer();
+                    deploymentMetaInfo = await Task.Run(() =>
+                        (DeploymentMetaInfo) serializer.Deserialize(deploymentInfoFile, typeof(DeploymentMetaInfo)));
```
Comment (f-porter) [inline Sources/ReleaseServer/ReleaseServer.WebApi/Services/FsReleaseArtifactService.cs:202]: It does not make sense to schedule a new task for this. Please make this synchronous.

File: Sources/ReleaseServer/ReleaseServer.WebApi.Test/Mappers/DeploymentMetaInfoMapperTest.cs [hunk 12]
```diff
@@ -1,30 +1,48 @@
...
+        
+        [Fact]
+        public void ConvertJsonToDeploymentMetaInfo_ByteArray()
+        {
+            var testMetaInfoByteArray = File.ReadAllBytes(Path.Combine(ProjectDirectory, "TestData", "testDeployment.json"));
+            
+            var parsedMeta = DeploymentMetaInfoMapper.ParseDeploymentMetaInfo(testMetaInfoByteArray);
 
-            var parsedMeta = DeploymentMetaInfoMapper.ParseDeploymentMetaInfo(Path.Combine(projectDirectory, "TestData","testDeployment.json"));
+            string test = null;
```
Comment (f-porter) [inline Sources/ReleaseServer/ReleaseServer.WebApi.Test/Mappers/DeploymentMetaInfoMapperTest.cs:43]: This variable is unused.

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs]: Yes, and an one of the invalid models is a missing `ReleaseDate` in the JSON that will be validated by `[JsonRequired]`.  

But no problem i will add a test with and empty 'DateTime' & a `DateTime` with an empty string:)

File: Sources/ReleaseServer/ReleaseServer.WebApi.Test/Services/ReleaseArtifactServiceTest.cs [hunk 13]
```diff
@@ -57,5 +56,137 @@ public async void TestGetLatestVersion_Not_Found()
...
+
+            var expectedValidationError = "the deployment meta information (deployment.json) is invalid!";
+
+            
+            //Act
+            var validationResult = await FsReleaseArtifactService.ValidateUploadPayload(testFormFile);
+
+            Assert.False(validationResult.IsValid);
+            Assert.Equal(expectedValidationError, validationResult.ValidationError);
+        }*/
```
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi.Test/Services/ReleaseArtifactServiceTest.cs:168]: These unit tests were refactored & fixed in the second PR.

File: Sources/ReleaseServer/ReleaseServer.WebApi/Mappers/DeploymentMetaMapper.cs (hunk 10 above)
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Mappers/DeploymentMetaMapper.cs]: You are right, It got removed in the second PR:)

File: Sources/ReleaseServer/ReleaseServer.WebApi/Services/FsReleaseArtifactService.cs (hunk 11 above)
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Services/FsReleaseArtifactService.cs:202]: Okay, i will do this

File: Sources/ReleaseServer/ReleaseServer.WebApi.Test/Mappers/DeploymentMetaInfoMapperTest.cs (hunk 12 above)
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi.Test/Mappers/DeploymentMetaInfoMapperTest.cs:43]: This test got removed in the second PR. 
I will remove it anyways:)

File: Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Extensions/DeploymentMetaInfoExtension.cs]: I added the unit tests in the second PR, the JsonSerializable<T> class got implemented, see: https://github.com/Traeger-GmbH/release-server/pull/59/commits/0a85c5876b45b04962e6ef40537f1f95f48e3327

In both cases, the deserializer recognized, that the ReleaseDate was not set.

File: Sources/ReleaseServer/ReleaseServer.WebApi.Test/Mappers/DeploymentMetaInfoMapperTest.cs (hunk 12 above)
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi.Test/Mappers/DeploymentMetaInfoMapperTest.cs]: Done, see https://github.com/Traeger-GmbH/release-server/pull/58/commits/f79a7b0384bf265984b2328cf1923beea7e9d023

File: Sources/ReleaseServer/ReleaseServer.WebApi/Services/FsReleaseArtifactService.cs (hunk 11 above)
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi/Services/FsReleaseArtifactService.cs]: ValidateUploadPayload() is now synchronous, see https://github.com/Traeger-GmbH/release-server/pull/58/commits/f79a7b0384bf265984b2328cf1923beea7e9d023

File: Sources/ReleaseServer/ReleaseServer.WebApi.Test/Services/ReleaseArtifactServiceTest.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (tiwalter) [reply Sources/ReleaseServer/ReleaseServer.WebApi.Test/Services/ReleaseArtifactServiceTest.cs]: Nevertheless, i uncommented and fixed these unit test also in this PR, see https://github.com/Traeger-GmbH/release-server/pull/58/commits/f79a7b0384bf265984b2328cf1923beea7e9d023

**Comment by f-porter** [comment]
> I think you missed the ReleaseArtifactModel when you removed the "Model" suffix.

Do you have overseen this comment? Did miss it or did you leave the suffix on purpose?

However this one seems to be left. When it is fixed I will merge this PR.

**Comment by tiwalter** [comment]
> 
> 
> > I think you missed the ReleaseArtifactModel when you removed the "Model" suffix.
> 
> Do you have overseen this comment? Did you miss it or did you leave the suffix on purpose?
> 
> However this one seems to be left. When it is fixed I will merge this PR.

Done, see https://github.com/Traeger-GmbH/release-server/pull/58/commits/379a4240a203beb8b391847d0ab5414aaf930963
